File: src/main/java/com/leet/medium/PathSumIII.go
package main

/**
 * 437. Path Sum III Medium
 *
 * https://leetcode.com/problems/path-sum-iii/
 *
 * Given the root of a binary tree and an integer targetSum, return the number
 * of paths where the sum of the values along the path equals targetSum.
 *
 * The path does not need to start or end at the root or a leaf, but it must go
 * downwards (i.e., traveling only from parent nodes to child nodes).
 *
 * Example 1:
 * Input: root = [10,5,-3,3,2,null,11,3,-2,null,1], targetSum = 8
 * Output: 3
 * Explanation: The paths that sum to 8 are shown.
 *
 * Example 2:
 * Input: root = [5,4,8,11,null,13,4,7,2,null,null,5,1], targetSum = 22
 * Output: 3
 *
 * Constraints:
 * The number of nodes in the tree is in the range [0, 1000]. -109 <= Node.val
 * <= 109 -1000 <= targetSum <= 1000
 */

type PathSumIII struct {
	targetSum int
}

func (this *PathSumIII) merge(arr, next []int, pos int, node *TreeNode) (int, int) {
	ptr, ans, compl := 0, 0, this.targetSum-node.Val
	for ptr < len(arr) {
		next[pos] = arr[ptr] + node.Val
		if arr[ptr] == compl {
			ans++
		}
		ptr++
		pos++
	}
	return pos, ans
}

func (this *PathSumIII) solve(node *TreeNode) ([]int, int) {
	if node == nil {
		return []int{}, 0
	}
	left, lans := this.solve(node.Left)
	right, rans := this.solve(node.Right)
	next := make([]int, len(left)+len(right)+1)
	var mergeLeftAns, mergeRightAns, pos int
	pos, mergeLeftAns = this.merge(left, next, 0, node)
	pos, mergeRightAns = this.merge(right, next, pos, node)
	ans := mergeLeftAns + mergeRightAns
	if node.Val == this.targetSum {
		ans++
	}
	next[pos] = node.Val
	return next, ans + lans + rans
}

func pathSum(root *TreeNode, targetSum int) int {
	sol := PathSumIII{targetSum}
	_, ans := sol.solve(root)
	return ans
}

File: src/main/java/com/leet/medium/PathSumIII.exs
defmodule Solution do
  def solve(node, target_sum) do
    if node == nil do
      {[], 0}
    else
      {left_arr, left_count} = solve(node.left, target_sum)
      {right_arr, right_count} = solve(node.right, target_sum)
      {arr, lcount} = merge(left_arr, [], node, target_sum)
      {arr, rcount} = merge(right_arr, arr, node, target_sum)
      count = if node.val == target_sum, do: 1 + lcount + rcount, else: lcount + rcount
      {[node.val | arr], count + left_count + right_count}
    end
  end

  @spec merge(arr :: [integer], ans :: [integer], node :: TreeNode.t(), target_sum :: integer) ::
          {[integer], integer}
  def merge(arr, ans, node, target_sum) do
    Enum.reduce(arr, {ans, 0}, fn e, acc ->
      {lis, count} = acc
      count = if e + node.val == target_sum, do: count + 1, else: count
      {[e + node.val | lis], count}
    end)
  end

  @spec path_sum(root :: TreeNode.t() | nil, target_sum :: integer) :: integer
  def path_sum(root, target_sum) do
    {_, count} = solve(root, target_sum)
    count
  end
end
